components/focus-trap/focus-trap.js

```diff
@@ -0,0 +1,97 @@
+import { css, html, LitElement } from 'lit-element/lit-element.js';
+import { getNextFocusable, getPreviousFocusable } from '../../helpers/focus.js';
+import { ifDefined } from 'lit-html/directives/if-defined.js';
+import { isComposedAncestor } from '../../helpers/dom.js';
+
+class FocusTrap extends LitElement {
+
+	static get properties() {
+		return {
+			trap: { type: Boolean, reflect: true }
+		};
+	}
+
+	static get styles() {
+		return css`
+			:host {
+				display: inline-block;
+			}
+		`;
+	}
+
+	constructor() {
+		super();
+		this.trap = false;
+		this._handleBodyFocus = this._handleBodyFocus.bind(this);
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+		document.body.addEventListener('focus', this._handleBodyFocus, true);
+	}
+
+	disconnectedCallback() {
+		super.disconnectedCallback();
+		document.body.removeEventListener('focus', this._handleBodyFocus, true);
+	}
+
+	focus() {
+		this.querySelector('.d2l-focus-trap-start').focus();
+	}
+
+	render() {
+		const tabindex = this.trap ? '0' : undefined;
+		return html`
+			<span class="d2l-focus-trap-start" @focusin="${this._handleStartFocusIn}" tabindex="${ifDefined(tabindex)}"></span>
+			<slot></slot>
+			<span class="d2l-focus-trap-end" @focusin="${this._handleEndFocusIn}" tabindex="${ifDefined(tabindex)}"></span>
+		`;
+	}
+
+	_focusFirst() {
+		const focusable = getNextFocusable(this.shadowRoot.querySelector('.d2l-focus-trap-start'));
+		if (focusable) focusable.focus();
```

Comment:
Is it possible in dialog's case that this could focus first on one element and then dialog would move it to another? More worried about the screen reader experience.

Reply:
Yep that would happen.  I did a test with VO and it seems to be ok... in the same way that landing on the trap-start or trap-end elements redirects wraps focus to the end or beginning and only the final active element is announced.  It's a valid concern though.  I _could_ add a property that enables consumers to opt-in/out of this behavior, but so far it seems unnecessary.

Reply:
Cool, seems fine then!